pkg/auth/login.go

```diff
@@ -0,0 +1,290 @@
+package auth
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+const loginPath = "/auth/login"
+const registerPath = "/auth/register"
+
+const (
+	loginUsingBrowser          = "Login with a web browser"
+	pasteAnAuthenticationToken = "Paste an authentication token"
+)
+
+type paramsOrError struct {
+	token        string
+	user         string
+	organization string
+	err          error
+}
+
+type Host struct {
+	Token        string `yaml:"token"`
+	User         string `yaml:"user"`
+	Organization string `yaml:"organization"`
+}
+
+var paramsOrErrorChan = make(chan paramsOrError)
+
+func HandleLogin(insightsHost string) error {
+	if _, err := os.Stat(polarisHostsFilepath); err == nil {
+		content, err := readPolarisHostsFile()
+		if err != nil {
+			return fmt.Errorf("reading polaris hosts file: %w", err)
+		}
+
+		if len(content) > 0 {
+			var reAuthenticate bool
+			err = survey.AskOne(&survey.Confirm{Message: fmt.Sprintf("You're already logged into %s. Do you want to re-authenticate?", insightsHost)}, &reAuthenticate)
+			if err != nil {
+				return fmt.Errorf("prompting re-authenticate: %w", err)
+			}
+			if !reAuthenticate {
+				// bail-out
+				return nil
+			}
+		}
+	}
+
+	selection := &survey.Select{
+		Message: "How would you like to authenticate Polaris?",
+		Options: []string{loginUsingBrowser, pasteAnAuthenticationToken},
+		Default: loginUsingBrowser,
+	}
+
+	var answer string
+	err := survey.AskOne(selection, &answer)
+	if err != nil {
+		return fmt.Errorf("asking how to authenticate: %w", err)
+	}
+
+	var user, token, organization string
+	if answer == loginUsingBrowser {
+		listener, err := net.Listen("tcp", ":0")
+		if err != nil {
+			panic(err)
+		}
+		localServerPort := listener.Addr().(*net.TCPAddr).Port
+		err = openBrowser(fmt.Sprintf(insightsHost + registerPath + "?source=polaris&callbackUrl=" + fmt.Sprintf("http://localhost:%d/auth/login/callback", localServerPort)))
+		if err != nil {
+			logrus.Fatal(err)
+		}
```

Comment:
If `openBrowser` fails, can we just print the URL to STDOUT?

Reply:
yea, that makes sense